### DISCUSSION TITLE
How to redefine a default implementation of an interface for a subclass?

### DISCUSSION THREAD

**Discussion by stephen-hlx** [raphw/byte-buddy#1620] [Q&A]
```java
public interface MyInterface {
  default void baseAction() {
    System.out.println("in default baseAction()");
  }
}

public class MyClass implements MyInterface {
  public void action() {
    System.out.println("in MyClass.action()");
  }
}
```

Is there a way for me to redefine the `baseAction` for `MyClass`?
Assuming `MyClass`'s constructor is unknown at this stage, or I cannot use `MyClass.getConstructor().newInstance()` to initialise it.
(the reason is that I only vend `MyInterface` and the client will be implementing their own subclasses like `MyClass` so I have no control how the instances of those subclasses are initialised)

**Comment by raphw** [comment]
You can redefine it with a Java agent. Have a look at `AgentBuilder` which allows you to change the code on a general basis. If you control the class loader, you can also redefine the class.

**Comment by stephen-hlx** [comment]
Thanks. Got caught up in something and i'll try and get back to this thread later.

**Comment by stephen-hlx** [comment]
Thanks Rafael.

(sorry, got caught up with some work and just got back to this)
I was reading into the `AgentBuilder`, but not sure if I fully understood.
I firstly want to make sure my requirement is actually achievable ...

Let me rephrase my requirement a bit:
I want to redefine the default implementation of the interface implemented by a class, and I also need to be able to perform the redefinition multiple times at runtime, and create different objects of the class. But these objects of the same class don't have to be used at the same time, like
1. redefine the base method for classA
2. initialise an object of classA, use it
3. redefine the base method for classA again, so the base method will do something else, different from that at step 1
4. initialise an object of classA, use it

The above need to have during runtime.

Sample code:
```java
interface TheBase {
  default void baseAction1() {}
  default void baseAction2(TheBase other) {}
}

class TheImpl1 implements TheBase {
  public void action() {}
}

class TheImpl2 implements TheBase {
  public void action() {}
}

class App {
  void run() {
    redefineTheImpl1_baseAction2_version1();
    TheImpl1 impl1_1 = new TheImpl1();
    TheImpl2 impl2_1 = new TheImpl2();
    impl1_1.baseAction2(impl2_1);

    redefineTheImpl1_baseAction2_version2();
    TheImpl1 impl1_2 = new TheImpl1();
    TheImpl2 impl2_2 = new TheImpl2();
    impl1_2.baseAction2(impl2_2);
  }

  void redefineTheImpl1_baseAction2_version1() {}
  void redefineTheImpl1_baseAction2_version2() {}
}
```

**Comment by stephen-hlx** [comment]
it turns out mockito already provides what i need, thanks!
